### PR TITLE
GEODE-8924: Add VM restart tests for Redis and Spring sessions

### PR DESCRIPTION
**Pull request by jdeppe-pivotal**
- Use Spring repositories to perform fetch and save operations.

Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [ ] Is your initial contribution a single, squashed commit?

- [ ] Does `gradlew build` run cleanly?

- [ ] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, check Concourse for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

File: geode-redis/src/distributedTest/java/org/apache/geode/redis/session/SessionsAndCrashesDUnitTest.java [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+    phase.set("CRASH 4 REBALANCING");
+    rebalanceAllRegions(server3);
+
+    phase.set("FINISHING");
+    GeodeAwaitility.await().during(10, TimeUnit.SECONDS).until(() -> true);
+
+    running.set(false);
+    future1.get();
+    future2.get();
+  }
```
Comment (nonbinaryprogrammer) [inline geode-redis/src/distributedTest/java/org/apache/geode/redis/session/SessionsAndCrashesDUnitTest.java:162]: Could you validate that the values on the session are correct, or at least not garbage? 
This could theoretically pass if the session updater threads were all stalled for the ten seconds the test is running for. We should confirm that all the attributes we expect attr-#-[0->NUM_SESSIONS] exist.
Even better would be to validate that the count at the end of the value is greater than or equal to the modCount at the end of the attribute name. This would indicate that we made it through more than one iteration and no data was lost in the rebalancing.

File: geode-redis/src/distributedTest/java/org/apache/geode/redis/session/SessionsAndCrashesDUnitTest.java [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+    Throwable latestException = null;
+
+    for (int i = 0; i < 10; i++) {
+      try {
+        sessionRepository.save(session);
+        return;
+      } catch (RedisSystemException rex) {
+        latestException = rex;
+        try {
+          Thread.sleep(2000);
```
Comment (nonbinaryprogrammer) [inline geode-redis/src/distributedTest/java/org/apache/geode/redis/session/SessionsAndCrashesDUnitTest.java:210]: why is this needed and can it be less than 2 seconds? If not then you might need to make the GeodeAwaitility.during() (line 157) at least 20 seconds, because it's possible that we could spend that long trying to save the session and end up not going through all 10 attempts.

File: geode-redis/src/distributedTest/java/org/apache/geode/redis/session/SessionsAndCrashesDUnitTest.java [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+    future1.get();
+    future2.get();
+  }
+
+  private Integer sessionUpdater(int index, AtomicBoolean running, AtomicReference<String> phase) {
+    int count = 0;
+    while (running.get()) {
+      int modCount = count % NUM_SESSIONS;
+      String sessionId = sessionIds.get(modCount);
+      Session session = findSession(sessionId);
```
Comment (nonbinaryprogrammer) [inline geode-redis/src/distributedTest/java/org/apache/geode/redis/session/SessionsAndCrashesDUnitTest.java:169]: between this and the call to saveSession(), the sessionUpdater method could take over 40 seconds. the wait in the test for the finish phase only waits 10 seconds, so a call to sessionUpdater could potentially be cut short.

**Review by nonbinaryprogrammer — CHANGES_REQUESTED** [review]

File: geode-redis/src/distributedTest/java/org/apache/geode/redis/session/SessionsAndCrashesDUnitTest.java [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+
+  @ClassRule
+  public static ExecutorServiceRule executor = new ExecutorServiceRule();
+
+  private static final int JEDIS_TIMEOUT =
+      Math.toIntExact(GeodeAwaitility.getTimeout().toMillis());
+  private static final int NUM_SESSIONS = 100;
+  private static final List<String> sessionIds = new ArrayList<>(NUM_SESSIONS);
+  private static MemberVM locator;
+  private static MemberVM server1;
```
[the remark this answers is not in the document]
Comment (jdeppe-pivotal) [reply geode-redis/src/distributedTest/java/org/apache/geode/redis/session/SessionsAndCrashesDUnitTest.java:64]: I want to have different servers to stop/start. `server1` is the lead member and thus has a much higher quorum weight. if it were to shut down then the remaining members would shut down since they would lose quorum. So instead, leave the lead member alone and use the other 2.

File: geode-redis/src/distributedTest/java/org/apache/geode/redis/session/SessionsAndCrashesDUnitTest.java [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+        .withConnectionToLocator(locatorPort));
+  }
+
+  @Before
+  @SuppressWarnings("unchecked")
+  public void setup() {
+    String[] args = new String[] {
+        "" + redisPorts[2],
+        "" + redisPorts[1],
+        "" + redisPorts[0]};
```
[the remark this answers is not in the document]
Comment (jdeppe-pivotal) [reply geode-redis/src/distributedTest/java/org/apache/geode/redis/session/SessionsAndCrashesDUnitTest.java:106]: I've removed the connection to server1 so that the connections are only established to the 2 servers that are to be shut down and restarted.

File: geode-redis/src/distributedTest/java/org/apache/geode/redis/session/SessionsAndCrashesDUnitTest.java (hunk 1 above)
Comment (jdeppe-pivotal) [reply geode-redis/src/distributedTest/java/org/apache/geode/redis/session/SessionsAndCrashesDUnitTest.java:182]: Added some validation.

File: geode-redis/src/distributedTest/java/org/apache/geode/redis/session/SessionsAndCrashesDUnitTest.java (hunk 2 above)
Comment (jdeppe-pivotal) [reply geode-redis/src/distributedTest/java/org/apache/geode/redis/session/SessionsAndCrashesDUnitTest.java]: Your comment prompted me to see if we can get away with just a single retry and it looks like the test is quite happy doing that.

File: geode-redis/src/distributedTest/java/org/apache/geode/redis/session/SessionsAndCrashesDUnitTest.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (nonbinaryprogrammer) [reply geode-redis/src/distributedTest/java/org/apache/geode/redis/session/SessionsAndCrashesDUnitTest.java:65]: 👍

**Review by nonbinaryprogrammer — APPROVED** [review]
looks good, thanks for addressing those comments!
